x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/apikey/CreateApiKeyRequest.java

```diff
@@ -144,6 +154,10 @@ public void setMetadata(Map<String, Object> metadata) {
 
     @Override
     public ActionRequestValidationException validate() {
+        if (Assertions.ENABLED && type == ApiKey.Type.CROSS_CLUSTER) {
```

Comment:
Wondering if it could make sense to use inheritance here: have an abstract base class and a CreateRestApiKeyRequest and a CreateCrossClusterAccessApiKeyRequest that override the type and validation. A lot of the validation steps below (like missing name) don't really make much sense for cross cluster key request.

Not a strong suggestion, just an intuition that although these are almost identical requests, they're still quite semantically distinct. LMKWYT.

Reply:
I didn't do it because it is a lot cascading changes. But it will help to make the code cleaner. I'll give it a go.

> A lot of the validation steps below (like missing name) don't really make much sense for cross cluster key request.

The name validation is still needed because the API takes the API Key's name from the user.

Reply:
This is done. I kept the existing `CreateApiKeyRequest` class name unchanged and created the base class `AbstractCreateApiKeyRequest` to avoid large cascading changes. The class has a builder `CreateApiKeyRequestBuilder` which would need to be renamed as well. I think we can leave the renames to a future pure-refactor PR so that this one is a bit easier to work with.

Reply:
> The name validation is still needed because the API takes the API Key's name from the user.

My bad, you're right. I got the API key name mixed up with fixed role name.